lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp

```diff
@@ -0,0 +1,135 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.h"
+
+#include <pxr/imaging/hd/tokens.h>
+#include <pxr/imaging/hd/sceneIndexPrimView.h>
+#include <pxr/imaging/hd/materialSchema.h>
+#include <pxr/imaging/hd/materialBindingSchema.h>
+#include <pxr/imaging/hd/materialBindingsSchema.h>
+#include <pxr/imaging/hd/containerDataSourceEditor.h>
+#include <pxr/imaging/hd/retainedDataSource.h>
+
+namespace FVP_NS_DEF {
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace{
+    static const TfToken purposes[] = { HdMaterialBindingsSchemaTokens->allPurpose };
+}
+
+// static
+DefaultMaterialSceneIndexRefPtr
+DefaultMaterialSceneIndex::New(
+    const HdSceneIndexBaseRefPtr &inputSceneIndex,
+    const PXR_NS::SdfPath& defaultMaterialPath,
+    const PXR_NS::SdfPathVector& defaultMaterialExclusionList)
+{    
+    return TfCreateRefPtr(
+        new DefaultMaterialSceneIndex(inputSceneIndex, defaultMaterialPath, defaultMaterialExclusionList));
+}
+
+DefaultMaterialSceneIndex::DefaultMaterialSceneIndex(
+    HdSceneIndexBaseRefPtr const &inputSceneIndex, 
+    const PXR_NS::SdfPath& defaultMaterialPath,
+    const PXR_NS::SdfPathVector& defaultMaterialExclusionList)
+  : HdSingleInputFilteringSceneIndexBase(inputSceneIndex), 
+    InputSceneIndexUtils(inputSceneIndex),
+    _defaultMaterialPath(defaultMaterialPath),
+    _defaultMaterialExclusionList(defaultMaterialExclusionList)
+{   
+}
+
+HdSceneIndexPrim DefaultMaterialSceneIndex::GetPrim(const SdfPath& primPath) const
+{
+    HdSceneIndexPrim prim = GetInputSceneIndex()->GetPrim(primPath);
+    if (! _isEnabled){
+        return prim;
+    }
+
+    _SetDefaultMaterial(prim);
+
+    return prim;
+}
+
+bool DefaultMaterialSceneIndex::_ShouldWeApplyTheDefaultMaterial(const HdSceneIndexPrim& prim)const
+{
+    // Only for meshes so far
+    if (HdPrimTypeTokens->mesh != prim.primType) {
+        return false;
+    }
+
+    // Check if it has a material which is not in the exclusion list
+    HdMaterialBindingsSchema bindings = HdMaterialBindingsSchema::GetFromParent(prim.dataSource);
+    HdMaterialBindingSchema  binding = bindings.GetMaterialBinding();
+    HdPathDataSourceHandle   bindingPathDS = binding.GetPath();
+    if (bindingPathDS) { // If a mesh prim has no material( bindingPathDS is empty), apply anyway the default material
+        const SdfPath materialPath = bindingPathDS->GetTypedValue(0.0f);
+        auto    foundExcludedMaterialPath = std::find(
```

Comment:
What size is this _defaultMaterialExclusionList expected to be?  We are looping over this for each prim.  If the list size is small, it's fine, but otherwise we can consider using an std::set.

Reply:
There is only one at this time, do you think I should switch to a std::set now ?

Reply:
No, just add a comment mentioning this as a possibility if the defaultMaterialExclusionList becomes large.